drivers/include/rn2xx3.h

```diff
@@ -0,0 +1,686 @@
+    RN2XX3_ERR_NO_FREE_CH,             /**< All channels are busy */
+    RN2XX3_ERR_SILENT,                 /**< Device is in Silent Immediately state */
+    RN2XX3_ERR_FR_CNT_REJOIN_NEEDED,   /**< Frame counter rolled over */
+    RN2XX3_ERR_BUSY,                   /**< MAC is not in Idle state */
+    RN2XX3_ERR_MAC_PAUSED,             /**< MAC was paused */
+    RN2XX3_ERR_INVALID_DATA_LENGTH,    /**< Wrong payload given */
+    RN2XX3_ERR_KEYS_NOT_INIT,          /**< Keys not configured ("mac join" command) */
+    RN2XX3_ERR_SLEEP_MODE,             /**< Failure because device is in sleep mode */
+    /* all other reply codes */
+    RN2XX3_REPLY_TX_MAC_OK,            /**< MAC transmission successful */
+    RN2XX3_REPLY_TX_MAC_ERR,           /**< MAC transmission failed */
+    RN2XX3_REPLY_TX_INVALID_DATA_LEN,  /**< Application payload too large */
+    RN2XX3_REPLY_TX_MAC_RX,            /**< Data received from server */
+    RN2XX3_REPLY_JOIN_ACCEPTED,        /**< Join procedure successful */
+    RN2XX3_REPLY_JOIN_DENIED,          /**< Join procedure failed */
+    RN2XX3_REPLY_TIMEOUT,              /**< No MAC reply received from server */
+    RN2XX3_REPLY_OTHER,                /**< Unknown reply */
+};
+
+/**
+ * @brief   Internal states of the device
+ */
+enum {
+    RN2XX3_INT_STATE_RESET,            /**< the device is being reset or initialized */
+    RN2XX3_INT_STATE_CMD,              /**< waiting command response */
+    RN2XX3_INT_STATE_IDLE,             /**< waiting for incoming commands */
+    RN2XX3_INT_STATE_SLEEP,            /**< sleep mode */
+    RN2XX3_INT_STATE_MAC_JOIN,         /**< waiting for mac join procedure reply */
+    RN2XX3_INT_STATE_MAC_TX,           /**< waiting for mac tx reply */
+    RN2XX3_INT_STATE_MAC_RX_PORT,      /**< waiting for mac rx port */
+    RN2XX3_INT_STATE_MAC_RX_MESSAGE,   /**< waiting for mac rx message */
+};
+
+/**
+ * @brief   LoRaMAC communication settings
+ */
+typedef struct {
+    uint32_t rx2_freq;                 /**< Center frequency used for second receive window */
+    loramac_dr_idx_t rx2_dr;           /**< Datarate used for second receive window */
+    uint8_t tx_port;                   /**< Application TX port (between 1 and 223) */
+    loramac_tx_mode_t tx_mode;         /**< TX mode, either confirmable of unconfirmable */
+    uint8_t rx_port;                   /**< RX port (between 1 and 223) */
+} loramac_settings_t;
+
+/**
+ * @brief   Configuration parameters for RN2483/RN2903 devices
+ */
+typedef struct {
+    uart_t uart;                       /**< UART interfaced the device is connected to */
+    uint32_t baudrate;                 /**< baudrate to use */
+    gpio_t pin_reset;                  /**< GPIO pin that is connected to the STATUS pin
+                                            set to GPIO_UNDEF if not used */
+} rn2xx3_params_t;
+
+/**
+ * @brief   RN2483/RN2903 device descriptor
+ */
+typedef struct {
```

Comment:
This struct is massive... And it wastes quite some memory on bad ordering. But I'm ok with improving later.